File: challenge_4/detect_xor.py
def xor_decrypt(ciphertext, key):
    plaintext = ""
    for byte in ciphertext:
        plaintext += chr(byte ^ key)
    return plaintext

def score_plaintext(plaintext):
    score = 0
    for char in plaintext:
        if char in "etaoin shrdlu":
            score += 1
    return score

def main():
    with open("data.txt", "r") as file:
        encoded_strings = [line.strip() for line in file.readlines()]

    potential_solutions = []

    for encoded_string in encoded_strings:
        for key in range(256):  # Try all possible single-byte keys
            plaintext = xor_decrypt(bytes.fromhex(encoded_string), key)
            score = score_plaintext(plaintext)
            potential_solutions.append((score, key, plaintext))

    # Sort potential solutions by score in descending order
    potential_solutions.sort(reverse=True)

    # Print the most likely plaintext
    best_score, best_key, best_plaintext = potential_solutions[0]
    print("Best Score:", best_score)
    print("Best Key:", hex(best_key))
    print("Best Plaintext:", best_plaintext)

if __name__ == "__main__":
    main()

File: challenge_4/detect_xor.cpp
#include <iostream>
#include <fstream>

int score_plaintext(std::string plaintext);
std::string xor_decrypt(std::string ciphertext, char key);

int main()
{
    std::fstream data("data.txt");

    if (!data.is_open()) // Check if file is open
    {
        std::cerr << "Error opeing file" << std::endl;
        return 1;
    }

    std::string ciphertext;
    std::string plaintext;
    std::string best_plaintext;

    int score;
    int best_score = -1000;
    
    char best_key = 0;
    
    // For each encoded string in the file
    while (std::getline(data, ciphertext))
    {

        // For each possible key
        for (char key = 0; key < 127; key++)
        {
            plaintext = xor_decrypt(ciphertext, key); // Decrypt the string using single-byte XOR
            score = score_plaintext(plaintext); // Score the resulting plaintext

            if (score > best_score) 
            {
                best_score = score;
                best_key = key; 
                best_plaintext = plaintext;
            }
        }
    }

    // Print the best result
    std::cout << "Best score: " << best_score << std::endl;
    std::cout << "Best key: " << best_key << std::endl;
    std::cout << "Best plaintext: " << best_plaintext << std::endl;
}

int score_plaintext(std::string plaintext)
{
    int score = 0;

    for (char c : plaintext)
    {
        if (isalpha(c) || c == 32)
            score++;
        else
            score--;
    }

    return score;
}

std::string xor_decrypt(std::string ciphertext, char key)
{
    std::string plaintext;
    char xored_byte;

    for (int i = 0; i < ciphertext.length() / 2; i++)
    {
        std::string hex_pair = ciphertext.substr(i * 2, 2); 
        xored_byte = std::stoi(hex_pair, nullptr, 16) ^ key; // Convert hex pair to int, xor with key
        plaintext += xored_byte;
    }

    return plaintext;
}
